### DISCUSSION TITLE
Lattice reference in calculating small-world

### DISCUSSION THREAD

**Discussion by qxu1994** [networkx/networkx#4754] [Q&A]
Hi, 

I was trying to understand how the lattice reference of a graph was established, but I encountered some issues that really confuse me. 

I first built a watts-strogratz small-world graph:
G = nx.watts_strogatz_graph(n = 10, k = 4, p = 0.5)

Theoretically, its lattice reference should look the same as the watts-strogratz graph when p equals to 0. And the lattice graph will look like a ring in which each node has connections with its four nearest neighbors:
G_lattice = nx.watts_strogatz_graph(n = 10, k = 4, p = 0)

However, I tested the lattice reference function of networkx using:
nx.lattice_reference(G)
The graph being output looks very different from G_lattice. Could I ask why they are different?

Thanks!

**Comment by dschult** [comment]
Setting `p=0` in a `watts_strogatz_graph` does create a lattice. 

The `lattice_reference` function is *not* intended to create a lattice. 
It is intended to perform "edge swaps" on a given graph to make it more lattice-like without changing the degree distribution.
So you start with a graph and perform `niter` edge swaps so long as the edge swaps make the adjacency matrix closer to a diagonal matrix. See the references for more details on what "close" to a diagonal matrix means. Notice that a lattice will have the non-zero entries close to the main diagonal if the nodes are ordered correctly.

**Reply by qxu1994** [reply]
Thanks, that makes sense! Could I ask whether it matters to keep the degree distribution identical? 

I'm curious about this because I saw literature papers mixed with whether the reference network should have the same degree distribution or just the average degree over all nodes (e.g., Steyvers & Tenenbaum, 2005; Dorogovtsev & Mendes, 2001). It seems that the Watts and Strogatz's (1998) paper did not control degree distribution either. 

Thanks again!

**Reply by dschult** [reply]
I don't know the latest literature consensus on whether to keep the degree distribution the same or not. I know that many algorithms do that.  Certainly maintaining the same average degree allows a **lot** of flexibility -- it's just the same number of edges!  So you could build a regular graph or a star graph and everything in between. :}

The `watts_strogatz_graph` follows their 1998 paper. But the small_world.py module and its functions follow slightly more recent literature. Edge swaps are pretty commonly employed as a way to look at a set of similar graphs well beyond the small world literature.

**Reply by qxu1994** [reply]
Thanks for the clarification. I really appreciate it! Since I'm working on graphs with thousands of nodes, the average degree one is particularly desirable in terms of the speed of calculation. But I'll be cautious about the potential difference between the two methods.

**Reply by dschult** [reply]
Edge swaps are not very costly in terms of calculation... Its much easier to keep the degree distribution the same than the number of triangles or something like that.  :}
